mpi4: implement hw guided and unguided MPI_Comm_split_type

hzhou opened the pull request:
## Pull Request Description

Fixes #4914
## Expected Impact

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -152,14 +152,52 @@ int MPIR_Comm_split_type_hw_guided(MPIR_Comm * comm_ptr, int key,
+    /* we'll try top-down until we hit upon a subset */
+    mpi_errno = MPIR_Comm_split_type_by_node(comm_ptr, key, &subcomm);
+    MPIR_ERR_CHECK(mpi_errno);
+
```
Is it OK that we always assume that the top level resource is node? E.g., maybe we want to start from network-cabinet (a group of nodes) level on a Dragonfly network. Do you think it might be necessary to make code more flexible?

hzhou replied:
We'll have to amend the code to support network topology, to be done in the future. Network topology will just be tried before `split_type_by_node`. So the code is not "inflexible", it just needs to be *extended*.

hzhou replied:
Added `TODO` comment.

minsii commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -152,14 +152,52 @@ int MPIR_Comm_split_type_hw_guided(MPIR_Comm * comm_ptr, int key,
+        mpi_errno = node_split(comm_ptr, key, topolist[i], &subcomm);
+        MPIR_ERR_CHECK(mpi_errno);
+
+        if (MPIR_Comm_size(subcomm) < orig_size) {
```
Maybe it is good to add a note that we return the first successful subcomm

hzhou replied:
Comment added

minsii replied:
I don't see comment added here in commit "comm: implment MPIR_Comm_split_type_hw_unguided". Did you add into a separate commit?

hzhou replied:
I thought the comment at the beginning of the function is sufficient: https://github.com/pmodels/mpich/blob/881ef456c47bac6a34ddce03e8d571104e95bb41/src/mpi/comm/comm_split_type.c#L165-L170

Let me add another short one here.

minsii commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -28,6 +28,9 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
+        /* NOTE: typical path goes down to device layer rather than here.
+         * Currently ch4 calls MPIR_Comm_split_type_by_node, which checks
+         * the "shmem_topo" infohints */
         mpi_errno = MPIR_Comm_split_type_self(comm_ptr, key, newcomm_ptr);
```
The note is a bit confusing to me. Does the typical path mean "MPI_Comm_split_type->MPIR_Comm_split_type_impl"?

I think it might be better to say that the MPIR approach for `TYPE_SHARED` is a fallback used only when the device layer does not handle such a split type.

hzhou replied:
Amended

minsii commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -55,6 +55,38 @@ int main(int argc, char *argv[])
+            MPI_Comm_size(comm, &newsize);
+            if (newsize < world_size) {
+                num_subsets++;
+            }
```
Can we report the `split_topo` name of failed subset?

minsii replied:
If the returned comm is not `COMM_NULL` but `newsize` >= `world_size`, do we need report it as an error?

hzhou replied:
It is allowed to return a comm with the same size as `wolrd_size`, for example, split by "node" and every process is on the same node. Returning either `COMM_NULL` or `world_size` is not strictly an error.

So it is tricky to do a good test here. Actually, if we do not test with `MPIR_CVAR_ODD_EVEN_CLIQUES=1`, we would probably always get split of world size (since typically we do not bind processes during tests).

But the tests without any validation seem useless, thus I take this approach that assumes some splits should end up with strictly sub-splits.

minsii replied:
I still don't get the idea why we check `num_subsets == 0` and post only a "warning" message for it (not error).

Maybe we could pass `mpiexec` argument to set the binding, and check the hint properly?

hzhou replied:
> I still don't get the idea why we check `num_subsets == 0` and post only a "warning" message for it (not error).

Oh, you mean I forgot to increment `errs`? I should do that.

> Maybe we could pass mpiexec argument to set the binding, and check the hint properly?

We'll need to fix the hardware platforms where tests are run to effectively test, and the functionality will depend on the `hwloc` support which only works well on Linux I believe.

hzhou replied:
Alright, I rewrote the tests (for `MPI_COMM_TYPE_HW_GUIDED`)  following the same tests as `shmem_topo` below.

minsii replied:
I think it might be better to report error "Expected comm size to be at most the world size" (not node size) in case we extend the implementation to support network topo.

hzhou commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -91,14 +91,65 @@ int MPIR_Comm_split_type_self(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
+        goto fn_exit;
+    }
+
+    int is_equal = 0;
```
I don't feel another name is any better. Let me just add a comment at both caller sites and the definition site for now.

minsii replied:
I don't see the comment added in commit "comm: implement MPIR_Comm_split_type_hw_guided"

hzhou replied:
Missed this one because it was added in the later commits. Now comment added.

hzhou replied:
Modified the code a bit to make sure we call `MPII_compare_info_hint` collectively.

hzhou commented on src/mpi/comm/commutil.c:
```diff
@@ -1092,3 +1092,58 @@ int MPIR_Comm_release_always(MPIR_Comm * comm_ptr)
+
+/* used by comm_split_type and its family */
+int MPII_compare_info_hint(const char *hint_str, MPIR_Comm * comm_ptr, int *info_args_are_equal)
+{
```
Added comment

hzhou commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -28,7 +28,13 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
     if (split_type == MPI_COMM_TYPE_SHARED) {
-        mpi_errno = MPIR_Comm_split_type_self(comm_ptr, split_type, key, newcomm_ptr);
+        mpi_errno = MPIR_Comm_split_type_self(comm_ptr, key, newcomm_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
```
Good catch. Fixed.

hzhou commented:
@minsii I believe all comments are addressed. Could you review it again? Since most changes are comments, I'll leave the testing later.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -0,0 +1,296 @@
+        MPIR_ERR_CHECK(mpi_errno);
+
+        if (MPIR_Comm_size(subcomm) < orig_size) {
+            /* found the first true sub-comm, return it */
```
Short comment added

hzhou commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -55,6 +55,67 @@ int main(int argc, char *argv[])
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    MPI_Info_free(&info);
```
New test added

minsii replied:
As now this test covers several different info keys, maybe we want to also print at which case the error occurs. E.g., "Testing MPI_COMM_TYPE_HW_GUIDED: expected MPI_COMM_NULL due to different info values."

minsii replied:
An additional note: the standard says "All MPI processes must provide the same value for the info key "mpi_hw_resource_type." but it does not explicitly say if we should return `COMM_NULL` with `MPI_SUCCESS` or an MPI error. I guess the intention is an error. Otherwise such a case should be listed under "MPI_COMM_NULL is also returned in newcomm in the following cases:"

hzhou commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -55,6 +55,67 @@ int main(int argc, char *argv[])
+        MPI_Comm_free(&comm);
+    }
+    if (rank % 2 == 0)
+        MPI_Info_free(&info);
```
new test added

minsii replied:
It looks like the standard allows one process to specify info but the others pass `INFO_NULL`. `COMM_NULL` is returned **only for the processes with `INFO_NULL`** (see quoted standard below). Thus, I assume that a valid newcomm should be returned for the processes passed a valid info.

>If an MPI process does not meet the above criteria, then MPI_COMM_NULL is returned in newcomm for such process.
MPI_COMM_NULL is also returned in newcomm in the following cases:
• No info key is provided.
• The info handle does not include the key "mpi_hw_resource_type" .

Besides, the tests covers only the INFO_NULL case, maybe we want to also add the cases for "info without key" and "info without mpi_hw_resource_type key"?

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
@minsii All comments addressed, plus some modifications to the tests. Could you review again?

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii reviewed (commented):
A few more comments for the test commit.

minsii commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -55,6 +55,67 @@ int main(int argc, char *argv[])
+        }
+        MPI_Comm_free(&comm);
+    }
 
```
Can we also add a test to check the returned info with `MPI_COMM_TYPE_HW_UNGUIDED`?

>If a valid info handle is provided as an argument, the MPI implementation sets the info key "mpi_hw_resource_t ype" for each MPI process in the group associated with returned newcomm communicator and the info key value is an implementation-deﬁned string that indicates the hardware resource type represented by newcomm

hzhou replied:
done.

minsii reviewed (commented):
Two additional comments

hzhou commented:
> It looks like the standard allows one process to specify info but the others pass INFO_NULL. COMM_NULL is returned only for the processes with INFO_NULL (see quoted standard below). Thus, I assume that a valid newcomm should be returned for the processes passed a valid info.

There is also this clause in the standard:
```
All MPI processes must provide the same value for the info key "mpi_hw_resource_type"
```

Thus some processes specify info and other does not, is incorrect. Thus I think returning `MPI_COMM_NULL` for *all* processes are appropriate.

Semantically it does make sense to return *partial* `MPI_COMM_NULL` unless user explicitly use `MPI_UNDEFINED` split_type.

hzhou commented:
> Besides, the tests covers only the INFO_NULL case, maybe we want to also add the cases for "info without key" and "info without mpi_hw_resource_type key"?

sure.

hzhou commented:
> An additional note: the standard says "All MPI processes must provide the same value for the info key "mpi_hw_resource_type." but it does not explicitly say if we should return COMM_NULL with MPI_SUCCESS or an MPI error. I guess the intention is an error. Otherwise such a case should be listed under "MPI_COMM_NULL is also returned in newcomm in the following cases:"

That sounds ok to me. However, from user point of view, I don't see much difference between "No info key is provided" or "info doesn't include the key "mpi_hw_resource_type", with "values of mpi_hw_resource_type are different". All these cases indicate a user error (as it contradicts to user intention).

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
>That sounds ok to me. However, from user's point of view, I don't see much difference between "No info key is provided" or "info doesn't include the key "mpi_hw_resource_type", with "values of mpi_hw_resource_type are different". All these cases indicate a user error (as it contradicts to user intention). So from user perspective, have two different error responses can be quite confusing.

I don't have a better idea. Maybe we can leave it for now and wait for a more precise error description in standard. E.g., new errors for MPI-4 features are not defined yet.

hzhou commented:
Re-reading the text, it says 
"If an MPI process does not meet the above criteria, then MPI_COMM_NULL is returned in newcomm for such process."

That means if the process is not pinned to e.g. `hwthread` and the user asks to split on `hwthread`, then we need return `MPI_COMM_NULL`. This is very different behavior from our current `shmem_topo` hints. This PR will need more work. In addition, the setting of info hints to the returned comm also need different logics than existing code.

hzhou commented:
> 
> 
> > That sounds ok to me. However, from user's point of view, I don't see much difference between "No info key is provided" or "info doesn't include the key "mpi_hw_resource_type", with "values of mpi_hw_resource_type are different". All these cases indicate a user error (as it contradicts to user intention). So from user perspective, have two different error responses can be quite confusing.
> 
> I don't have a better idea. Maybe we can leave it for now and wait for a more precise error description in standard. E.g., new errors for MPI-4 features are not defined yet.

On second thought, the `MPI_COMM_NULL` cases that `No info key is provided` and `info does not include mpi_hw_resource_type` probably both refer to when *all* processes are providing no info or no info key, i.e. it is still consistent. With that interpretation, we should return error if *some* processes provide info and *others* don't.

minsii commented:
>Re-reading the text, it says
"If an MPI process does not meet the above criteria, then MPI_COMM_NULL is returned in newcomm for such process."
This may mean that if the process is not pinned to e.g. hwthread and the user asks to split on hwthread, then we need to return MPI_COMM_NULL. If this interpretation is correct, then it is very different behavior from our current shmem_topo hints. This PR will need more work. In addition, the setting of info hints to the returned comm also need different logics than existing code.

I am not very sure what changes are needed in the PR. I split the above text into two points. Can you please let me know if my understanding matches yours? And which part is missing in this PR?
1. If the topo mapping fails then we need return COMM_NULL (thus, we need check whether the mapping fails). I can guess two cases: (1) the user asks for a XXX mapping but such topology does not exist or is not supported by the MPI 
implementation. (2) such mapping is possible but the mapping operation fails. In both cases, we need return COMM_NULL.

2. If some processes map successfully but some others fail, COMM_NULL should be returned only on the failed processes (not on all processes)

Point-2 is also confirmed with the standard text below:
>The MPI implementation will return in the group of the output communicator newcomm the largest subset of MPI processes that match the splitting criterion.

hzhou commented:
> 
> 
> > Re-reading the text, it says
> > "If an MPI process does not meet the above criteria, then MPI_COMM_NULL is returned in newcomm for such process."
> > This may mean that if the process is not pinned to e.g. hwthread and the user asks to split on hwthread, then we need to return MPI_COMM_NULL. If this interpretation is correct, then it is very different behavior from our current shmem_topo hints. This PR will need more work. In addition, the setting of info hints to the returned comm also need different logics than existing code.
> 
> I am not very sure what changes are needed in the PR. I split the above text into two points. Can you please let me know if my understanding matches yours? And which part is missing in this PR?
> 
>     1. If the topo mapping fails then we need return COMM_NULL (thus, we need check whether the mapping fails). I can guess two cases: (1) the user asks for a XXX mapping but such topology does not exist or is not supported by the MPI
>        implementation. (2) such mapping is possible but the mapping operation fails. In both cases, we need return COMM_NULL.
> 
>     2. If some processes map successfully but some others fail, COMM_NULL should be returned only on the failed processes (not on all processes)
> 
> 
> Point-2 is also confirmed with the standard text below:
> 
> > The MPI implementation will return in the group of the output communicator newcomm the largest subset of MPI processes that match the splitting criterion.

The particular criterion in the text is "Each output communicator newcomm corresponds to a *single* instance of the specified hardware resource type." So if we have 4 processes on the node *not binded* to any *single* core, and the user ask to split the comm by "core", we'll have to return `MPI_COMM_NULL`. The current `node_split` will happily return a comm include all 4 processes, each does not corresponds to a single instance (a single core). So we'll need extend the current hwtopo utilities to add the "single instance" detection functionality.

hzhou commented:
Actually, I am not quite sure what is the correct way of interpreting the text. @raffenet Maybe you have a better understanding (regarding https://github.com/pmodels/mpich/pull/5071#issuecomment-782754242)

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
@minsii I have made sure that MPI_COMM_TYPE_HW_GUIDED will always split into groups that consist a single instance of a given type. In addition, I also added a split hint of `bindset`, which will split based on each process's bindset. This is useful when, for example, the user allocates some process on some cores and the rest on the other cores.

I believe your other comments on the tests are also addressed now.

The PR is rebased and tests clean. Could you review it one more time?

minsii commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -175,14 +175,65 @@ int MPIR_Comm_split_type_hw_guided(MPIR_Comm * comm_ptr, int key,
-                                     MPIR_Info * info_ptr, MPIR_Comm ** newcomm_ptr)
+int MPIR_Comm_split_type_hw_unguided(MPIR_Comm * comm_ptr, int key, MPIR_Info * info_ptr,
+                                     MPIR_Comm ** newcomm_ptr)
 {
```
Can we squash the above format/renaming with the previous stub commit?

minsii commented on src/util/mpir_hwtopo.c:
```diff
@@ -462,6 +462,24 @@ MPIR_hwtopo_gid_t MPIR_hwtopo_get_obj_by_name(const char *name)
             gid = HWTOPO_GET_GID(class, non_io_ancestor->depth, non_io_ancestor->logical_index);
         }
+    } else if (!strcmp(name, "bindset")) {
+        char buf[100];          /* covers up to 800 PUs */
```
Can you please explain the semantics of `bindset`? We might want to have a detailed description in the commit message. Could you please also update the MPICH wiki for process binding?

minsii replied:
The "PU" unit info is lost in "bindset" naming. Maybe we want to rename it to something like "puset"?. Actually the "set" format seems suitable for all resource types. E.g., define set of CPUs for each rank.

It looks like the "set" format is not currently supported by hydra `-bind-to` option.  How do you deal with it for the binding operation? I'do see a commit in this PR

Not sure whether it is needed to have "bindset" in the initial impl. I feel that we might need more work to make it complete. Maybe we want to drop it from this PR for now?

hzhou replied:
The commit is dropped, but I'll reply anyway for completeness --

> The "PU" unit info is lost in "bindset" naming. Maybe we want to rename it to something like "puset"?. Actually the "set" format seems suitable for all resource types. E.g., define set of CPUs for each rank.
> 

Probably "cpuset" is a more common name. However, I want to emphasize the "bind" part. Cpuset by itself doesn't mean much. Maybe "cpubindset"?

> It looks like the "set" format is not currently supported by hydra `-bind-to` option. How do you deal with it for the binding operation? I'do see a commit in this PR
> 
It's supported, e.g. `mpiexec -bind-to -user:0,1+2,1+2`.  This PR https://github.com/pmodels/mpich/pull/5095 further extents the syntax so it now supports `mpiexec -bind-to -user:0:1-63x63`

minsii commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -55,6 +55,63 @@ int main(int argc, char *argv[])
+        int newsize;
+        MPI_Comm_size(comm, &newsize);
+        if (!(newsize < world_size)) {
+            printf("MPI_COMM_TYPE_HW_UNGUIDED: Expected comm to be a proper sub communicator\n");
```
Can you please explain why `new_size == world_size` is wrong? And maybe it is better to use `newsize >= world_size` rather than `!(<)`.

hzhou replied:
`MPI_COMM_TYPE_HW_UNGUIDED` needs to return *strict* subset of original comm per specification, so it requires the new comm size strictly less than the original size. `newsize >= world_size` just read silly since how can the comm split resulting in a bigger comm than original?

minsii replied:
Do you want to keep the "!(newsize < world_size)" code?

minsii replied:
I am on the other side. Anyway I don't think this line is worth for us to spend more time on it. I am fine with it

minsii commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -55,6 +55,63 @@ int main(int argc, char *argv[])
+            errs++;
+        }
+        char resource_type[100] = "";
+        int has_key = 0;
```
It might be safer to define `char resource_type[MPI_MAX_INFO_VAL]` as the test can be used for non-MPICH implementations.

minsii replied:
To ensure portability with other MPI implementations, I think setting buffer size be `MPI_MAX_INFO_VAL` is the safest way. I think that is why standard defines it, right?

minsii replied:
Any response for this comment?

hzhou commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -91,14 +91,88 @@ int MPIR_Comm_split_type_self(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
+        *newcomm_ptr = node_comm;
+        node_comm = NULL;
+        goto fn_exit;
+    }
```
>     * Would you mind describing the diffs between `MPIR_Comm_split_type_hw_guided ` and current `shmem_topo ` in the commit message? I feel that it will help followers understand the changed logic

(I'll address other two points separately ). I think now that we have official HW_GUIDED, the `shmem_topo` hints feature should get deleted. I am planning to do so right after we merge this PR. That's why I don't think we need too much comments on `shmem_topo`. What do you think?

minsii replied:
>  I think now that we have official HW_GUIDED, the shmem_topo hints feature should get deleted. I am planning to do so right after we merge this PR. That's why I don't think we need too much comments on shmem_topo. What do you think?

Sounds good to me

hzhou commented:
@minsii I have addressed your comments. Could you give it another review?

minsii commented on src/mpi/errhan/errnames.txt:
```diff
@@ -378,6 +378,8 @@ be in the range 0 to %d
 **infonkey:Requested nth key does not exist
 **infonkey %d %d:Requested key %d but this MPI_Info only has %d keys
+**infonoteq:MPI_Info are not consistent for all processes
+**infonoteq %s:MPI_Info key %s is not consistent for all processes
```
A minor comment here: maybe we want to keep the `infonoteq` consistent and more specific. E.g.,
```
**infonoteq: The MPI_Info key value are not consistent for all processes
**infonoteq %s: The MPI_Info value with key %s is not consistent for all processes
```

hzhou replied:
Updated as you suggested.

minsii commented on src/util/mpir_hwtopo.c:
```diff
@@ -462,6 +462,24 @@ MPIR_hwtopo_gid_t MPIR_hwtopo_get_obj_by_name(const char *name)
+            }
+        }
+        HASH_VALUE(buf, j, gid);
     } else
```
The whole piece of code for "bindset" is not easy to understand even after reading the description in commit message.  I can see that you use bitmap + hash value to obtain a valid `gid` for each PU-set. what does up to 800 PUs mean? Is 800 the limit for the number of PUs per set?

But please see my above comment. If it is not necessary to push the bindset together with this PR, I'd suggest we drop it for now.

hzhou reviewed (commented):
Oh, sorry, forgot to push these comments

hzhou commented:
@minsii Your comments addressed (plus the old comments I forgot to push :) ). Could you review it again?

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/tcp
